File: lib/engine/anomalyDetector.ts
import type {
  MonthlyDataPoint,
  ControlChartParams,
  AnomalyResult,
  IndicatorStatus,
  Direction,
  DataNature,
} from '../types';
import { selectChartType, computeControlChartParams, detectControlChartAnomalies } from './controlChart';
import { detectMonthlyChanges } from './monthlyChange';
import { detectPeerDeviation } from './peerComparison';

export interface AnalysisResult {
  status: IndicatorStatus;
  anomalies: AnomalyResult[];
  controlChart: ControlChartParams | null;
}

/**
 * 三重異常偵測引擎 — 整合管制圖、月增減、同儕比較
 * @param skipControlChart 若為 true，跳過管制圖計算與異常偵測（如經營管理類指標）
 */
export function analyzeIndicator(
  monthlyData: MonthlyDataPoint[],
  peerValue: number | null,
  direction: Direction,
  dataNature: DataNature = 'continuous',
  skipControlChart: boolean = false,
): AnalysisResult {
  const anomalies: AnomalyResult[] = [];

  // 排序數據點（按年月遞增）
  const sorted = [...monthlyData]
    .filter(dp => dp.value !== null)
    .sort((a, b) => (a.year * 12 + a.month) - (b.year * 12 + b.month));

  if (sorted.length === 0) {
    return { status: 'neutral', anomalies: [], controlChart: null };
  }

  // 機制一：管制圖（用最近 24 個月計算管制限，全部數據做異常偵測）
  let controlChart: ControlChartParams | null = null;
  if (!skipControlChart) {
    const recent24 = sorted.slice(-24);
    const chartType = selectChartType(recent24, dataNature);
    controlChart = computeControlChartParams(recent24, chartType);
    if (controlChart) {
      const ccAnomalies = detectControlChartAnomalies(sorted, controlChart, direction);
      anomalies.push(...ccAnomalies);
    }
  }

  // 機制二：月增減幅度
  const monthlyAnomalies = detectMonthlyChanges(sorted, direction);
  anomalies.push(...monthlyAnomalies);

  // 機制三：同儕值比較（對最新數據點）
  const latest = sorted[sorted.length - 1];
  if (latest.value !== null && peerValue !== null) {
    const peerAnomaly = detectPeerDeviation(
      latest.value,
      peerValue,
      direction,
      latest.year,
      latest.month
    );
    if (peerAnomaly) {
      anomalies.push(peerAnomaly);
    }
  }

  // 綜合判定
  const status = resolveStatus(anomalies, latest);

  return { status, anomalies, controlChart };
}

/**
 * 綜合判定等級
 *
 * 優先級：
 * 🔴 Alert    — 管制圖 Rule 1（3σ 超限）
 * 🟠 Warning  — 管制圖 Rule 2-5，或 月增減不利 + 同儕比較也不利
 * 🟡 Watch    — 僅月增減不利 或 僅同儕比較不利
 * 🟢 Good     — 無任何異常
 * 🔵 Excellent — 改善方向且優於同儕
 */
function resolveStatus(
  anomalies: AnomalyResult[],
  latest: MonthlyDataPoint | undefined
): IndicatorStatus {
  if (!latest || latest.value === null) return 'neutral';

  // 只看最新月份相關的異常
  const allRelevant = anomalies.filter(a =>
    (a.year === undefined && a.month === undefined) ||
    (a.year === latest.year && a.month === latest.month)
  );

  const unfavorable = allRelevant.filter(a => a.direction === 'unfavorable');
  const favorable = allRelevant.filter(a => a.direction === 'favorable');

  // 檢查最嚴重的不利異常
  const hasAlert = unfavorable.some(a => a.severity === 'alert');
  const hasWarning = unfavorable.some(a => a.severity === 'warning');
  const hasWatch = unfavorable.some(a => a.severity === 'watch');

  // 多重不利因素加重
  const unfavorableMechanisms = new Set(unfavorable.map(a => a.mechanism));
  const multipleUnfavorable = unfavorableMechanisms.size >= 2;

  if (hasAlert) return 'alert';
  if (hasWarning) return 'warning';
  if (multipleUnfavorable && hasWatch) return 'warning'; // 兩種以上 watch 升級為 warning
  if (hasWatch) return 'watch';

  // 無不利異常，檢查是否卓越
  const hasExcellent = favorable.some(a => a.severity === 'excellent');
  const favorableMechanisms = new Set(favorable.map(a => a.mechanism));
  if (hasExcellent && favorableMechanisms.size >= 2) return 'excellent';

  return 'good';
}

/**
 * 計算特定月份的狀態（用於狀態矩陣）
 */
export function computeMonthStatus(
  allData: MonthlyDataPoint[],
  targetYear: number,
  targetMonth: number,
  peerValue: number | null,
  direction: Direction,
  controlChart: ControlChartParams | null
): IndicatorStatus {
  const sorted = [...allData]
    .filter(dp => dp.value !== null)
    .sort((a, b) => (a.year * 12 + a.month) - (b.year * 12 + b.month));

  const targetPoint = sorted.find(dp => dp.year === targetYear && dp.month === targetMonth);
  if (!targetPoint || targetPoint.value === null) return 'neutral';

  const anomalies: AnomalyResult[] = [];

  // 管制圖規則（對這個特定點）
  if (controlChart) {
    const hasSigma = controlChart.sigma > 0;
    const hasVarLimits = controlChart.variableLimits && controlChart.variableLimits.length > 0;

    if (hasSigma || hasVarLimits) {
      const value = targetPoint.value;

      // 取此點的管制限
      const vl = controlChart.variableLimits?.find(
        l => l.year === targetYear && l.month === targetMonth
      );
      const ucl = vl?.ucl ?? controlChart.ucl;
      const lcl = vl?.lcl ?? controlChart.lcl;
      const ucl2 = vl?.ucl2 ?? controlChart.ucl2;
      const lcl2 = vl?.lcl2 ?? controlChart.lcl2;

      // Rule 1
      if (value > ucl && (direction === 'lower' || direction === 'monitor')) {
        anomalies.push({ mechanism: 'control_chart', rule: 'rule1', severity: 'alert', direction: 'unfavorable', message: '超出 3σ 管制上限 (失控)', value, referenceValue: ucl, year: targetYear, month: targetMonth });
      }
      if (value < lcl && lcl > 0 && (direction === 'higher' || direction === 'monitor')) {
        anomalies.push({ mechanism: 'control_chart', rule: 'rule1', severity: 'alert', direction: 'unfavorable', message: '低於 3σ 管制下限 (失控)', value, referenceValue: lcl, year: targetYear, month: targetMonth });
      }

      // Rule 2
      if (value > ucl2 && value <= ucl && (direction === 'lower' || direction === 'monitor')) {
        anomalies.push({ mechanism: 'control_chart', rule: 'rule2', severity: 'warning', direction: 'unfavorable', message: '超出 2σ 警戒線', value, referenceValue: ucl2, year: targetYear, month: targetMonth });
      }
      if (value < lcl2 && value >= lcl && lcl2 > 0 && (direction === 'higher' || direction === 'monitor')) {
        anomalies.push({ mechanism: 'control_chart', rule: 'rule2', severity: 'warning', direction: 'unfavorable', message: '低於 2σ 警戒線', value, referenceValue: lcl2, year: targetYear, month: targetMonth });
      }

      // Favorable checks
      if (value > ucl && direction === 'higher') {
        anomalies.push({ mechanism: 'control_chart', severity: 'excellent', direction: 'favorable', message: '顯著高於 3σ 管制上限', value, referenceValue: ucl, year: targetYear, month: targetMonth });
      }
      if (value < lcl && lcl > 0 && direction === 'lower') {
        anomalies.push({ mechanism: 'control_chart', severity: 'excellent', direction: 'favorable', message: '顯著低於 3σ 管制下限', value, referenceValue: lcl, year: targetYear, month: targetMonth });
      }
    }
  }

  // 月增減
  const targetIdx = sorted.findIndex(dp => dp.year === targetYear && dp.month === targetMonth);
  if (targetIdx > 0) {
    const prev = sorted[targetIdx - 1];
    if (prev.value !== null && prev.value !== 0) {
      const changeRate = (targetPoint.value - prev.value) / Math.abs(prev.value);
      if (Math.abs(changeRate) >= 0.10) {
        const isIncrease = changeRate > 0;
        let dir: 'unfavorable' | 'favorable' = 'unfavorable';
        if (direction === 'lower') dir = isIncrease ? 'unfavorable' : 'favorable';
        else if (direction === 'higher') dir = isIncrease ? 'favorable' : 'unfavorable';

        anomalies.push({
          mechanism: 'monthly_change',
          severity: dir === 'unfavorable' ? 'watch' : 'excellent',
          direction: dir,
          message: `月增減 ${(changeRate * 100).toFixed(1)}%`,
          value: targetPoint.value,
          referenceValue: prev.value,
          year: targetYear,
          month: targetMonth,
        });
      }
    }
  }

  // 同儕比較
  if (peerValue !== null) {
    const peerAnomaly = detectPeerDeviation(
      targetPoint.value,
      peerValue,
      direction,
      targetYear,
      targetMonth
    );
    if (peerAnomaly) anomalies.push(peerAnomaly);
  }

  // 判定
  const unfavorable = anomalies.filter(a => a.direction === 'unfavorable');
  const favorable = anomalies.filter(a => a.direction === 'favorable');

  if (unfavorable.some(a => a.severity === 'alert')) return 'alert';
  if (unfavorable.some(a => a.severity === 'warning')) return 'warning';
  const mechCount = new Set(unfavorable.map(a => a.mechanism)).size;
  if (mechCount >= 2) return 'warning';
  if (unfavorable.some(a => a.severity === 'watch')) return 'watch';
  if (favorable.length >= 2) return 'excellent';
  return 'good';
}

File: backend/apps/analysis/services/anomaly_detector.py
"""
三重異常偵測引擎 — 整合管制圖、月增減、同儕比較

狀態判定優先級：
🔴 Alert    — 管制圖 Rule 1（3σ 超限）
🟠 Warning  — 管制圖 Rule 2-5，或多重不利因素
🟡 Watch    — 僅月增減不利 或 僅同儕比較不利
🟢 Good     — 無任何異常
🔵 Excellent — 多重改善訊號 + 優於同儕
⚪ Neutral  — 資料不足
"""
from __future__ import annotations

from dataclasses import dataclass

from .control_chart import (
    AnomalyResult,
    ControlChartParams,
    MonthlyDataPoint,
    compute_control_chart_params,
    detect_control_chart_anomalies,
    select_chart_type,
)
from .monthly_change import detect_monthly_changes
from .peer_comparison import detect_peer_deviation


@dataclass
class AnalysisResult:
    status: str  # IndicatorStatus
    anomalies: list[AnomalyResult]
    control_chart: ControlChartParams | None


def analyze_indicator(
    monthly_data: list[MonthlyDataPoint],
    peer_value: float | None,
    direction: str,
    data_nature: str = "continuous",
    skip_control_chart: bool = False,
    target_value: float | None = None,
) -> AnalysisResult:
    """三重異常偵測引擎

    target_value: 啟用挑戰平均值模式時，由呼叫端傳入目標值（顯示單位：% / ‰ / 原值）。
    """
    anomalies: list[AnomalyResult] = []

    sorted_data = sorted(
        [dp for dp in monthly_data if dp.value is not None],
        key=lambda dp: dp.year * 12 + dp.month,
    )

    if not sorted_data:
        return AnalysisResult(status="neutral", anomalies=[], control_chart=None)

    # Mechanism 1: Control chart
    control_chart: ControlChartParams | None = None
    if not skip_control_chart:
        recent_25 = sorted_data[-25:]
        chart_type = select_chart_type(recent_25, data_nature)
        control_chart = compute_control_chart_params(recent_25, chart_type, target_value=target_value)
        if control_chart:
            cc_anomalies = detect_control_chart_anomalies(sorted_data, control_chart, direction)
            anomalies.extend(cc_anomalies)

    # Mechanism 2: Monthly change
    monthly_anomalies = detect_monthly_changes(sorted_data, direction)
    anomalies.extend(monthly_anomalies)

    # Mechanism 3: Peer comparison
    latest = sorted_data[-1]
    if latest.value is not None and peer_value is not None:
        peer_anomaly = detect_peer_deviation(
            latest.value, peer_value, direction, latest.year, latest.month
        )
        if peer_anomaly:
            anomalies.append(peer_anomaly)

    status = _resolve_status(anomalies, latest)
    return AnalysisResult(status=status, anomalies=anomalies, control_chart=control_chart)


def _resolve_status(anomalies: list[AnomalyResult], latest: MonthlyDataPoint | None) -> str:
    if not latest or latest.value is None:
        return "neutral"

    all_relevant = [
        a for a in anomalies
        if (a.year is None and a.month is None)
        or (a.year == latest.year and a.month == latest.month)
    ]

    unfavorable = [a for a in all_relevant if a.direction == "unfavorable"]
    favorable = [a for a in all_relevant if a.direction == "favorable"]

    # 與 dashboard_bulk view 統一：watch 機制再多也不升級為 warning。
    # 升級只發生在 anomaly 本身就是 warning/alert 級別時（如管制圖 rule1/rule2）。
    has_alert = any(a.severity == "alert" for a in unfavorable)
    has_warning = any(a.severity == "warning" for a in unfavorable)
    has_watch = any(a.severity == "watch" for a in unfavorable)

    if has_alert:
        return "alert"
    if has_warning:
        return "warning"
    if has_watch:
        return "watch"

    has_excellent = any(a.severity == "excellent" for a in favorable)
    favorable_mechanisms = {a.mechanism for a in favorable}
    if has_excellent and len(favorable_mechanisms) >= 2:
        return "excellent"

    return "good"
